docs/rfc/007-endpoint-errors.md

```diff
@@ -0,0 +1,201 @@
+  probably just be `RuntimeException`s. This in turn might have knock-on effects in the codebase depending on the exact
+  implementation used.
+* Introducing further types (whether as checked exceptions or otherwise) that determine control flow may result in them 
+  spreading throughout a code base and result in further coupling to the RPC layer. As an example, using checked exceptions
+  could encourage the consumer to throw the exception upwards rather than dealing with it at the callsite.
+  
+### Mandatory vs. Optional
+
+As discussed above, declared errors can be either made to be optional, or mandatory. Both versions come with their own
+pros and cons, and picking one side or the other inevitably involves making sacrifices.
+
+**The case for mandatory handling**
+
+_Forcing_ consumers to handle your exceptional cases is beneficial to the _producer_. You can have confidence that
+consumers will be encouraged to handle the exceptional cases, even if they are not reading your documentation. Acknowledging
+the presence of exceptional cases will be required at the minimum.
+
+There are benefits as a consumer as well - additional errors, behaviors, or edge cases will be exposed to you on upgrading
+your client, and you should (in theory) never miss out on a new case (assuming you keep your client relatively up to date).
+
+**The case for optional handling**
+
+Optional handling favors the _consumer_ - existing code and behaviors will not have to change at the outset, and they
+can opt into improved error handling as necessary. The obvious downside is that producers could publish new errors until
+the cows come home, but lazy consumers may never actually handle these edge cases.
+
+## Implementation
+
+Generally, the client-side implementation should be done using whatever language features are best supported by the language
+itself. In Java, we have decided that that is checked exceptions - but in a language such as TypeScript that might be
+something akin to union types.
+
+### Compatibility
+
+It is worth noting that the changes suggested in this RFC do not entail a wire-format change, and as such should not
+affect backwards (or forwards) compatibility. Old servers that throw exceptions will continue to use `SerializableError`,
+and clients that do not upgrade can continue to catch `SerializableError`s as they did before.
+
+The intention instead is to better leverage Conjure types to provide better tooling for categorising errors into their
+types when they are received, and to improve documentation around what errors can be thrown by an endpoint.
+
+#### Backwards Compatibility of Error Fields
+
+As part of upgrading errors to a first-class part of the API, we should additionally treat errors in a first-class way
+when it comes to backwards compatibility. As the new suggested tooling will make errors more usable, accessing fields
+will also be more important, and (just as with regular fields) removing them will become more problematic.
```

Comment:
This gets into some of the issues we original described as out of scope, in particular all parameters are currently to-stringed by the java server implementations.

I agree that error data should be backwards compatible in the long term, but for the sake of this RFC I'm not confident we need to specify this. We may need to relax the constraint in order to support proper structured parameters in the future. I'd considered making all java a generated accessors for error fields optional such that deserialization may fail on individual fields without impacting the overall meaning of the error.